isolation_test.go

```diff
@@ -0,0 +1,250 @@
+package tidb_test
+
+import (
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/mockstore"
+	"github.com/pingcap/tidb/store/mockstore/mocktikv"
+	"github.com/pingcap/tidb/util/testkit"
+	"github.com/pingcap/tidb/util/testleak"
+)
+
+var _ = Suite(&testIsolationSuite{})
+
+type testIsolationSuite struct {
+	cluster   *mocktikv.Cluster
+	mvccStore *mocktikv.MvccStore
+	store     kv.Storage
+	dom       *domain.Domain
+}
+
+func (s *testIsolationSuite) SetUpSuite(c *C) {
+	testleak.BeforeTest()
+	s.cluster = mocktikv.NewCluster()
+	mocktikv.BootstrapWithSingleStore(s.cluster)
+	s.mvccStore = mocktikv.NewMvccStore()
+	store, err := mockstore.NewMockTikvStore(
+		mockstore.WithCluster(s.cluster),
+		mockstore.WithMVCCStore(s.mvccStore),
+	)
+	c.Assert(err, IsNil)
+	s.store = store
+	tidb.SetSchemaLease(0)
+	tidb.SetStatsLease(0)
+	s.dom, err = tidb.BootstrapSession(s.store)
+	c.Assert(err, IsNil)
+	tidb.SetCommitRetryLimit(0)
+}
+
+func (s *testIsolationSuite) TearDownSuite(c *C) {
+	tidb.SetCommitRetryLimit(10)
+	s.dom.Close()
+	s.store.Close()
+	testleak.AfterTest(c)()
+}
+
+/*
+These test cases come from the paper <A Critique of ANSI SQL Isolation Levels>.
+The sign 'P0', 'P1'.... can be found in the paper. These cases will run under snapshot isolation.
+*/
+func (s *testIsolationSuite) TestP0DirtyWrite(c *C) {
+	session1 := testkit.NewTestKitWithInit(c, s.store)
+	session2 := testkit.NewTestKitWithInit(c, s.store)
+
+	session1.MustExec("drop table if exists x;")
+	session1.MustExec("create table x (id int primary key, c int);")
+	session1.MustExec("insert into x values(1, 1);")
+
+	session1.MustExec("begin;")
+	session1.MustExec("update x set c = c+1 where id = 1;")
+	session2.MustExec("begin;")
+	session2.MustExec("update x set c = c+1 where id = 1;")
+	session1.MustExec("commit;")
+	_, err := session2.Exec("commit;")
+	c.Assert(err, NotNil)
+}
+
+func (s *testIsolationSuite) TestP1DirtyRead(c *C) {
+	session1 := testkit.NewTestKitWithInit(c, s.store)
+	session2 := testkit.NewTestKitWithInit(c, s.store)
+
+	session1.MustExec("drop table if exists x;")
+	session1.MustExec("create table x (id int primary key, c int);")
+	session1.MustExec("insert into x values(1, 1);")
+
+	session1.MustExec("begin;")
+	session1.MustExec("update x set c = c+1 where id = 1;")
+	session2.MustExec("begin;")
+	session2.MustQuery("select c from x where id = 1;").Check(testkit.Rows("1"))
+	session1.MustExec("commit;")
+	session2.MustExec("commit;")
+}
+
+func (s *testIsolationSuite) TestP2NonRepeatableRead(c *C) {
```

Comment:
Maybe I was wrong, but from my understanding, this is "Read Skew", not "NonRepeatableRead" ?

Reply:
these cases are similar and confused, sometimes the queries are same except the order. but I make sure that all cases come from the paper<A Critique of ANSI SQL Isolation Levels>.